numpy/core/tests/test_ufunc.py

```diff
@@ -111,67 +111,76 @@ def test_generic_loops(self):
         just looked at the signatures registered in the build directory to find
         relevant functions.
 
-        Fixme, currently untested:
-
-            PyUFunc_ff_f_As_dd_d
-            PyUFunc_FF_F_As_DD_D
-            PyUFunc_f_f_As_d_d
-            PyUFunc_F_F_As_D_D
-            PyUFunc_On_Om
-
         """
         fone = np.exp
         ftwo = lambda x, y: x**y
         fone_val = 1
         ftwo_val = 1
+
+        # check unary PyUFunc_f_f_As_d_d
+        msg = "PyUFunc_f_f_As_d_d"
+        x = np.full(10, np.single(0), dtype=np.double)
+        assert_almost_equal(fone(x), fone_val, err_msg=msg)
+        # check unary PyUFunc_F_F_As_D_D
+        msg = "PyUFunc_F_F_As_D_D"
+        x = np.full(10, np.csingle(0), dtype=np.cdouble)
+        assert_almost_equal(fone(x), fone_val, err_msg=msg)
         # check unary PyUFunc_f_f.
         msg = "PyUFunc_f_f"
-        x = np.zeros(10, dtype=np.single)[0::2]
+        x = np.zeros(10, dtype=np.single)
```

Comment:
Have not checked carefully, but I am confused by the removal of the `[::2]`. This probably has a purpose here to ensure a non-contiguous array, I think; which is important to test that SIMD optimization doesn't take a wrong path.

Reply:
Thanks for bringing up ``[0::2]`` part.

I'm not sure if what ``[0::2]`` is supposed to test. The non-contiguous array is plausible, but even if it is, the test still past without it - so whatever ``[::2]`` tests is *probably* not being caught by the tests. Even if it is, would you expect SIMD to change the *dtypes* of the input or output arrays?

I'll add ``[::2]`` back in iff you think it's the better idea, but then I would like to find an instance where a test actually fails without it.

Reply:
This does check equality of arrays/values on top of equality of data types,  no? I see a lot of assert almost equal.

Reply:
It tests ``abs(desired-actual) < 1.5 * 10**(-decimal)``. Looking at the (docs)[https://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.assert_almost_equal.html] there is actually a warning to avoid using this in favour of ``assert_allclose``. Perhaps I should change these?

Reply:
And no ``assert_almost_equal`` does **not** check the arrays datatype